spectec/spec/8-reduction.watsup

```diff
@@ -158,26 +158,29 @@ rule Step_read/call:
   z; (CALL x)  ~>  (REF.FUNC_ADDR $funcaddr(z)[x]) (CALL_REF)
 
 rule Step_read/call_ref-null:
-  z; (REF.NULL ht) (CALL_REF x)  ~>  TRAP
+  z; (REF.NULL ht) (CALL_REF x?)  ~>  TRAP
 
 rule Step_read/call_ref-func:
-  z; val^n (REF.FUNC_ADDR a) (CALL_REF x)  ~>  (FRAME_ m `{f} (LABEL_ m `{epsilon} instr*))
+  z; val^n (REF.FUNC_ADDR a) (CALL_REF x?)  ~>  (FRAME_ m `{f} (LABEL_ m `{epsilon} instr*))
   -- if $funcinst(z)[a] = fi
   -- Expand: fi.TYPE ~~ FUNC (t_1^n -> t_2^m)
-  -- if fi.CODE = FUNC x (LOCAL t)* (instr*)
+  -- if fi.CODE = FUNC x' (LOCAL t)* (instr*)
   -- if f = {LOCAL val^n ($default(t))*, MODULE fi.MODULE}
 
 
 rule Step_read/return_call:
   z; (RETURN_CALL x)  ~>  (REF.FUNC_ADDR $funcaddr(z)[x]) (RETURN_CALL_REF)
 
-rule Step_read/return_call_ref-frame:
-  z; (FRAME_ k `{f} val'* val^n ref (RETURN_CALL_REF x) instr*)  ~>  val^n ref (CALL_REF x)
+
+rule Step_read/return_call_ref-frame-null:
+  z; (FRAME_ k `{f} val* (REF.NULL ht) (RETURN_CALL_REF x?) instr*)  ~>  TRAP
```

Comment:
```suggestion
  z; (REF.NULL ht) (RETURN_CALL_REF x?)  ~>  TRAP
```

Reply:
A potential issue with this version is that it results in non-deterministic step, due to 
```
rule Step_read/return_call_ref-label:
  z; (LABEL_ k `{instr'*} val* (RETURN_CALL_REF x?) instr*)  ~>  val* (RETURN_CALL_REF x?)
```

Rewriting this rule as
```
rule Step_read/return_call_ref-label:
  z; (LABEL_ k `{instr'*} val* (REF.FUNC_ADDR a) (RETURN_CALL_REF x?) instr*)  ~>  val* (REF.FUNC_ADDR) (RETURN_CALL_REF x?)
```
would remove the nondeterminism, but now this rule becomes a bit lengthy.

So there would be three options:
1) choose concise `-label` rule and lengthy `-null` rule
2) choose concise `-null` rule and lengthy `-label` rule
3) choose concise, nondeterministic rules

What do you think?

Reply:
Ah, good point. In that case, perhaps keep the rule as, i.e. (1).